frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -38,25 +184,28 @@ const sendMessageToClient = async function (payload) {
     client.postMessage(payload)
   }
 }
+*/
 
 self.addEventListener('message', function (event) {
-  console.debug(`[sw] message from ${event.source.id}. Current store:`, store)
+  console.debug(`[sw] message from ${event.source.id} of type ${event.data?.type}.`)
   // const client = await self.clients.get(event.source.id)
   // const client = await self.clients.get(event.clientId)
   if (typeof event.data === 'object' && event.data.type) {
     console.debug('[sw] event received:', event.data)
     switch (event.data.type) {
-      case 'set':
-        store[event.data.key] = event.data.value
-        break
-      case 'get':
-        event.source.postMessage({
-          response: store[event.data.key]
+      case 'sbp': {
+        const port = event.data.port
+        ;(async () => await sbp(...deserializer(event.data.data)))().then((r) => {
```

Comment:
Should any consideration be given here regarding possible malicious code calling selectors that shouldn't be called?

Reply:
Do you have some specific scenarios in mind? In terms of malicious code provenance, I see two options: XSS or similar and a hostile contributor.

For XSS, I've proposed how we should get rid of DOM Purify and similar because sanitising is fundamentally more insecure than other approaches (although I'm not aware of any XSS vulnerabilities in the app). Implementing a CSP and sandboxing contracts would further restrict the avenues for smuggling code.

For hostile contributions, it's a tougher problem to solve, although if we did have checks here and someone changed them, they'd possibly stand out. 

However, I'm not sure it'd be easy to retrofit some filtering here. I'd say that _probably_ we could disallow calling 'private' selectors without ill effects, but security-wise, the app _needs_ to interact with contracts to function, and I think malicious actions look just like regular actions. For example, one could exfiltrate data by sending it as a DM. However, this is fundamentally indistinguishable from someone creating a DM.

Furthermore, even if we _did_ implement filtering, it can easily be bypassed by performing those actions directly. For example, a would-be attacker could just call the APIs directly (granted, we could implement filtering also in the `fetch` event handler).

Because of all of this, I don't think we can do much here. However, we could look into ways of using the Cache API creatively. For example, we could look into ways of filtering out 'secretKeys' from the browsing context and making them SW only.